File: packages/watchdog/src/rps-scanner.ts
const RPS_PATTERNS = [
  'ignore all previous instructions',
  'you are now',
  'new instructions',
  'disregard your training',
  'pretend you are',
  'override safety',
  '忽略所有之前的指令',
  '你现在是一个',
  '新指令',
  '无视你的训练',
  '假装你是',
  '覆盖安全',
]

const RPS_DISABLE_THRESHOLD = 3

export function scanRPS(
  text: string,
  _fieldName: 'prompt' | 'description',
  context?: { rpsConsecutiveFailures: number },
): { detected: boolean; patterns: string[]; skipped?: boolean } {
  if (context && isRPSDisabled(context)) {
    return { detected: false, patterns: [], skipped: true }
  }
  if (!text || text.length === 0) {
    return { detected: false, patterns: [] }
  }
  const lower = text.toLowerCase()
  const matched = RPS_PATTERNS.filter(p => lower.includes(p.toLowerCase()))
  return { detected: matched.length > 0, patterns: matched }
}

export function isRPSDisabled(state: { rpsConsecutiveFailures: number }): boolean {
  return state.rpsConsecutiveFailures >= RPS_DISABLE_THRESHOLD
}

export function resetRPSFailureCounter(state: { rpsConsecutiveFailures: number }): void {
  state.rpsConsecutiveFailures = 0
}

File: intervention/src/rps_scanner.py
"""RPSScanner — detects risky prompt suggestions via 12 regex patterns (6 EN + 6 CN)."""
import re
from dataclasses import dataclass
from typing import List


@dataclass
class RPSMatch:
    pattern_id: str
    pattern: str
    location: int
    severity_tier: str


_PATTERNS = [
    ("en-1", "ignore all previous instructions", "warn"),
    ("en-2", "you are now", "warn"),
    ("en-3", "new instructions", "warn"),
    ("en-4", "disregard your training", "warn"),
    ("en-5", "pretend you are", "warn"),
    ("en-6", "override safety", "warn"),
    ("cn-1", "忽略所有之前的指令", "warn"),
    ("cn-2", "你现在是一个", "warn"),
    ("cn-3", "新指令", "warn"),
    ("cn-4", "无视你的训练", "warn"),
    ("cn-5", "假装你是", "warn"),
    ("cn-6", "覆盖安全", "warn"),
]


class RPSScanner:
    def scan(self, prompt: str, description: str = "") -> List[RPSMatch]:
        if not prompt and not description:
            return []
        matches: List[RPSMatch] = []
        for pattern_id, pattern_text, tier in _PATTERNS:
            for source_text in (prompt, description):
                if not source_text:
                    continue
                idx = source_text.lower().find(pattern_text.lower())
                if idx >= 0:
                    matches.append(RPSMatch(
                        pattern_id=pattern_id,
                        pattern=pattern_text,
                        location=idx,
                        severity_tier=tier,
                    ))
                    break
        return matches
